### ISSUE TITLE
Multi WAN + Captive Portal not working

### ISSUE THREAD

**Issue by MYSense** [opnsense/core#1166]
Guys,

I am running on OPNsense 16.1.18-i386. I am not able to get Captive Portal to work on a Multi-WAN scenario. Captive Portal works fine if there is no policy based routing applied to the default LAN rule. As soon as I set a Gateway Group, Captive Portal fails to show, though I still get internet.

The 'ipfw' rules fail to apply when gateway group is set.

I have posted in the forum as well. Please find the details in the link below:
https://forum.opnsense.org/index.php?topic=3600.0

**Comment by L1ghtn1ng** [comment]
You should be running 16.7 and see if you have that issue on that version as 16.1 is not supported anymore

**Comment by srijannandi** [comment]
Upgraded to OPNsense 16.7.3-i386, still facing the same issue.

**Comment by AdSchellevis** [comment]
@srijannandi just tried it on my end and can confirm it doesn't work. Looks like a kernel issue. Found this online from some time ago https://lists.freebsd.org/pipermail/freebsd-pf/2010-July/005739.html , not sure if its related, there doesn't seem to be a quick fix.

**Comment by srijannandi** [comment]
Thanks, AD.

Do we have a sysctl setting to specify the order such as these:

sysctl net.inet.ip.pfil

net.inet.ip.pfil.inbound=pf, ipfw*
net.inet.ip.pfil.outbound=pf, ipfw*

Maybe this can keep the order fixed.

**Comment by AdSchellevis** [comment]
@srijannandi No (its not part of FreeBSD), but the loading order should be handled correctly by the rc script https://github.com/opnsense/core/blob/master/src/etc/rc.ipfw

**Comment by srijannandi** [comment]
AD,

Is the correct order PF and then IPFW?

Because, if I set default as gateway in the LAN to any rule. I have ipfw process the packet first which then pass it to pf. That's how the captive portal appears for authentication. Once authentication is done, it goes to pf for processing.

As soon as I set gateway group in LAN, this breaks and ipfw does not come into picture, and pf sends the packets out.

**Comment by AdSchellevis** [comment]
yes, the current order is correct (ipfw loaded before pf), its likely a bug in combination with route-to. 
Using ipfw in combination with pf sometimes provides provides some headaches (like the loading order).
In case you want to dig in the code, these two areas might be of interest:

https://github.com/opnsense/src/blob/master/sys/netpfil/pf/pf.c#L5459
https://github.com/opnsense/src/blob/master/sys/netpfil/pf/pf.c#L5434

(call if_output directly like mentioned in https://lists.freebsd.org/pipermail/freebsd-pf/2010-July/005739.html)

You can find some additional information about the normal package flow in this blog post in case your interested http://deadloop.blogspot.nl/2010/04/who-comes-first-ipfw-or-pf.html

Currently I don't have a lot of time to really dig into this.

**Comment by AdSchellevis** [comment]
Just performed some additional testing, the problem is that the pfil ordering can't be correct for both dummynet (traffic shaper) and route-to / captive portal.

pfSense has added a kernel hack to adjust the pfil ordering, we do the same in our scripting. 
To handle the traffic shaping correctly, we need to catch the traffic before NAT happens, which is what https://github.com/opnsense/core/blob/master/src/etc/rc.ipfw#L39 does.

If we change the order, route-to/Captive portal works (easy to test, just unload ipfw and load it again), but the shaper won't be able to match all the traffic correctly (due to being post nat).

In case you want to test the same I did, execute the following steps on your console (after captive portal has been started):


```
ipfw flush
kldunload dummynet
kldload ipfw

```

_note: you need to add a rule to accept traffic without policy based routing, so your internal machine can access the internal webserver._


Long story short, without serious kernel chances there is no simple fix here. We could add an advanced option to change the order somehow, but it would be difficult for people to understand what it actually does.

**Comment by fichtner** [comment]
I think a fix in FreeBSD will be sensible and possible due to the fact that pf(4) just doesn't adhere to packet filter chaining during route-to. The ipfw equivalent seems to work, which is used by the captive portal to redirect, isn't that right?

**Comment by fichtner** [comment]
working patch available, test builds soon

**Comment by fichtner** [comment]
Call for testing instructions here: https://forum.opnsense.org/index.php?topic=3704.0

**Comment by ra1nb0w** [comment]
with 	OPNsense 17.7.1_2-amd64 I see the same problem and I can't find 83fd8a6 on master.
I lost something?
thanks

**Comment by fichtner** [comment]
This is old. If you have trouble with Multi-WAN + Captive Portal, enable "Shared forwarding" under Firewall: Settings: Advanced.

**Comment by ra1nb0w** [comment]
ops. thank for the hit. have a nice day
